lib/streamlit/server/Server.py

```diff
@@ -429,52 +428,64 @@ def add_preheated_report_session(self):
         This is used to start running the user's script even before the first
         browser connects.
         """
-        session = self._add_browser_connection(PREHEATED_REPORT_SESSION)
+        session = self._create_report_session(PREHEATED_REPORT_SESSION)
         session.handle_rerun_script_request(is_preheat=True)
 
-    def _add_browser_connection(self, ws):
-        """Register a connected browser with the server
+    def _create_report_session(self, ws):
+        """Register a connected browser with the server.
 
         Parameters
         ----------
-        ws : _BrowserWebSocketHandler or PREHEATED_REPORT_CONTEXT
-            The newly-connected websocket handler
+        ws : _BrowserWebSocketHandler or PREHEATED_REPORT_SESSION
+            The newly-connected websocket handler.
 
         Returns
         -------
         ReportSession
-            The ReportSession associated with this browser connection
+            The newly-created ReportSession for this browser connection.
 
         """
-        if ws not in self._session_infos:
+        if PREHEATED_REPORT_SESSION in self._session_info_by_id:
+            assert len(self._session_info_by_id) == 1
+            LOGGER.debug("Reusing preheated context for ws %s", ws)
+            session = self._session_info_by_id[PREHEATED_REPORT_SESSION].session
+            del self._session_info_by_id[PREHEATED_REPORT_SESSION]
```

Comment:
I wonder if all this could be put in a function. It seems this is a `get_preheaded_report_session`.